package/MDAnalysis/analysis/bat.py

```diff
@@ -0,0 +1,422 @@
+This module was adapted from AlGDock [Minh2020]_.
+
+
+See Also
+--------
+:class:`~MDAnalysis.analysis.dihedrals.Dihedral`
+   class to calculate dihedral angles for a given set of atoms or residues
+:func:`MDAnalysis.lib.distances.calc_dihedrals()`
+   function to calculate dihedral angles from atom positions
+
+
+Example applications
+--------------------
+
+The :class:`~MDAnalysis.analysis.bat.BAT` class defines bond-angle-torsion
+coordinates based on the topology of an atom group and interconverts between
+Cartesian and BAT coordinate systems. For example, we can determine internal
+coordinates for residues 5-10 of adenylate kinase (AdK). The trajectory is
+included within the test data files::
+
+   import MDAnalysis as mda
+   from MDAnalysisTests.datafiles import PSF, DCD
+   import numpy as np
+
+   u = mda.Universe(PSF, DCD)
+
+   # selection of atomgroups
+   selected_residues = u.select_atoms("resid 5-10")
+
+   from MDAnalysis.analysis.bat import BAT
+   R = BAT(selected_residues)
+
+   # Calculate BAT coordinates for a trajectory
+   R.run()
+
+   # Reconstruct Cartesian coordinates from BAT coordinates
+   bat = R.bat[0]
+   XYZ = R.Cartesian(bat)
+
+   # The difference between the original and reconstructed coordinates
+   # should be zero.
+   print(np.sum(np.abs(XYZ - selected_residues.positions)>1E-6))
+
+After R.run(), the coordinates can be accessed with :attr:`R.bat`.
+
+
+References
+----------
+
+.. [Chang2003] Chang, Chia-En, Michael J Potter, and Michael K Gilson (2003).
+   "Calculation of Molecular Configuration Integrals". *Journal of Physical
+   Chemistry B* 107(4): 1048–55. doi:`10.1021/jp027149c
+   <https://doi.org/10.1021/jp027149c>`_
+
+.. [Hikiri2016] Hikiri, Simon, Takashi Yoshidome, and Mitsunori Ikeguchi (2016).
+   "Computational Methods for Configurational Entropy Using Internal and
+   Cartesian Coordinates." *Journal of Chemical Theory and Computation*
+   12(12): 5990–6000. doi:`10.1021/acs.jctc.6b00563
+   <https://doi.org/10.1021/acs.jctc.6b00563>`_
+
+.. [Minh2020] Minh, David D L (2020). "Alchemical Grid Dock (AlGDock): Binding
+   Free Energy Calculations between Flexible Ligands and Rigid Receptors."
+   *Journal of Computational Chemistry* 41(7): 715–30.
+   doi:`10.1002/jcc.26036 <https://doi.org/10.1002/jcc.26036>`_
+
+"""
+from __future__ import absolute_import
+# from six.moves import zip, range
+
+import numpy as np
+
+import warnings
+
+import MDAnalysis as mda
+from MDAnalysis.analysis.base import AnalysisBase
+
+from MDAnalysis.lib.distances import calc_bonds, calc_angles, calc_dihedrals
+
+
+def _sort_atoms_by_mass(atoms, reverse=False):
+    r""" Sorts a list of atoms by name and then by mass
+
+    Parameters
+    ----------
+    ag_o : list of Atoms
+      List to sort
+    reverse : bool
+      Atoms will be in descending order of mass
+
+    Returns
+    -------
+    ag_n : list of Atoms
+      Sorted list
+    """
+    return sorted(sorted(atoms, key = lambda atom:atom.name), \
```

Comment:
With this double sort, is the idea to sort by mass then by name as a tiebreaker?  If so I think better is to do `lambda x: (x.mass, x.name)`.  Tuples will get sorted by their first item, then second etc

Reply:
Yes, that is the idea. I want the ordering to be reproducible and was unaware of this trick of sorting by tuples. I am switching the atom name to atom index.